### PR TITLE
Implement AW11 CP

### PR DESCRIPTION
**Pull request by Klas-Norling**
# Implement aw11 in RABE repo
Aims to solve : [issue number you worked on](link_to_the_issue_you_worked_on)

## Status of pull request
- [x] Full functionality
- [x] Example file in python_examples
- [x] Removed env files
- [ ] Code format restrictions

## What is your solution?
Mapping of the rabe scheme aw11 to python.

## How did you test your solution?
Tested all the functions implemented and compiled and build with maturin.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,4 +8,5 @@ target
 .settings
 .project
 rabe.iml
-Cargo.lock
\ No newline at end of file
+Cargo.lock
+*.env
```
[the remark this answers is not in the document]
Comment (Leohemmingsson) [reply .gitignore:12]: Seems like I had wrong, might not be able to ignore if it is not added here?

File: rabe_py/src/aw11/types.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            Err(e) => return Err(PyErr::new::<PyValueError, _>(format!("{}", e))),
+        };
+        Ok(value)
+    }
+}
+
+
+
+
+/// An AW11 Public Key (pk) and vector containing Public Key
```
Comment (ivario123) [inline rabe_py/src/aw11/types.rs:50]: end all sentences with a .

File: rabe_py/src/aw11/types.rs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        let value = match serde_json::to_string(&self.pks) {
+            Ok(value) => value,
+            Err(e) => return Err(PyErr::new::<PyValueError, _>(format!("{}", e))),
+        };
+        Ok(value)
+    }
+}
+
+
+/// An AW11 Master Key (msk)
```
Comment (ivario123) [inline rabe_py/src/aw11/types.rs:82]: Maybe we should link to the classes in the doc comments
```rust
/// A wrapper for [`AW11 Master Key`](rabe::schemes::aw11:Aw11MasterKey)
/// ---
/// Used to interaface with python.

File: rabe_py/src/aw11/types.rs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+#[pyclass]
+pub struct PyAw11MasterKey {
```
Comment (ivario123) [inline rabe_py/src/aw11/types.rs:84]: Not sure if our wrapper classes should be named
```Rust
pub struct PyAw11...
// Or
pub struct Aw11...
```
I think that the python users have no idea that it is a rust wrapper, so adding the extra py... might be a cause of confusion.

Comment (ivario123) [reply rabe_py/src/aw11/types.rs:84]: If we change to 
```rust
pub struct Aw11...
```
then we need to change the import structure to be
```rust
use rabe::schemes::aw11::{
    Aw11GlobalKey as RustAw11GlobalKey, // Or similar
...
...
};
```

File: rabe_py/src/lib.rs [hunk 5]
```diff
@@ -1,14 +1,19 @@
-use pyo3::prelude::*;
+//! Code Formatting constrictions
```
Comment (ivario123) [inline rabe_py/src/lib.rs:1]: Change this doc comment to be something more meningful like
```rust
//! rabe_py
//! ---
//! 
//!  Allows for python interfacing with the [`rabe`](rabe) crate.
//!  So far, we provide interfaces for 
//! - Aw11
//! 
//! We plan on supporting all schemes provided by [`rabe::schemes`](rabe::schemes).
//! 
//! ## Examples
//!
//! ### Aw11
//! Some good aw11 example, wrapped in
//! ```python
//! ...
//! ...
//! ```
//!

**Review by ivario123 — COMMENTED** [review]
In general it looks good, the file documentation comments needs some work though, try to open the documentation as it stands currently
```bash
cargo doc --open
```

File: rabe_py/python_examples/testAW11.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
+# AW11 Scheme and examples implemented in python which use the functions
```
Comment (ivario123) [inline rabe_py/python_examples/testAW11.py:1]: Put this in a seperate folder
```
python_examples/aw11/aw11.py
```
Or just rename it to
```
python_examples/aw11.py
```

File: rabe_py/src/aw11/types.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
+//! AW11 implementation of structs and types
```
[the remark this answers is not in the document]
Comment (Klas-Norling) [reply rabe_py/src/aw11/types.rs]: Fixed

File: rabe_py/src/lib.rs [hunk 8]
```diff
@@ -1,14 +1,15 @@
-use pyo3::prelude::*;
 
-/// Formats the sum of two numbers as string.
-#[pyfunction]
-fn sum_as_string(a: usize, b: usize) -> PyResult<String> {
-    Ok((a + b).to_string())
-}
+
```
[the remark this answers is not in the document]
Comment (Klas-Norling) [reply rabe_py/src/lib.rs:7]: Fixed

File: .gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (Klas-Norling) [reply .gitignore]: Fixed

File: rabe_py/src/aw11/mod.rs [hunk 9]
```diff
@@ -0,0 +1,131 @@
+//! AW11 Encryption and Decryption interface for python
+//! ---
+//! 
+//! This is based on the  [original paper](https://some webage) and is implented in [the aw11 source](rabe::schemes::aw11)
```
[the remark this answers is not in the document]
Comment (Klas-Norling) [reply rabe_py/src/aw11/mod.rs]: Fixed

File: rabe_py/src/aw11/mod.rs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        Ok(plaintext) => plaintext,
+        Err(e) => return Err(PyErr::new::<PyValueError, _>(format!("{}", e))),
+    };
+    Ok(plaintext)
+}
+
+///A function that adds the classes and functions that should be accessible from python
+#[pymodule]
+pub fn aw11(_py: Python, m: &PyModule) -> PyResult<()> {
+    m.add_function(wrap_pyfunction!(setup, m)?)?;
```
[the remark this answers is not in the document]
Comment (Klas-Norling) [reply rabe_py/src/aw11/mod.rs]: Fixed

File: rabe_py/src/aw11/types.rs [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/// An AW11 Secret Key (sk)
+#[pyclass]
+pub struct PyAw11SecretKey {
+    pub sk: Aw11SecretKey,
+}
+
+#[pymethods]
+impl PyAw11SecretKey {
```
[the remark this answers is not in the document]
Comment (Klas-Norling) [reply rabe_py/src/aw11/types.rs]: Fixed

File: rabe_py/python_examples/testAW11.py (hunk 6 above)
Comment (Klas-Norling) [reply rabe_py/python_examples/testAW11.py:1]: Fixed

**Pull request merged**